subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc

```diff
@@ -0,0 +1,95 @@
+
+The example task `generateDocs` generates documentation for a project in the form of HTML pages. The task should be organized underneath the bucket `Documentation`. The description should express its intent.
+
+++++
+<sample id="taskGroupDescription" dir="userguide/bestPractices/taskGroupDescription" title="A task declaring the group and description">
+    <sourcefile file="build.gradle"/>
+</sample>
+++++
+
+The output of the tasks report reflects the assigned values.
+
+----
+$ gradle tasks
+
+> Task :tasks
+
+...
+
+Documentation tasks
+-------------------
+generateDocs - Generates the HTML documentation for this project.
+
+...
+----
+
+[[sec:avoid_use_of_imperative_logic]]
+=== Avoid the use of imperative logic
+
+The Gradle runtime does not enforce a specific style for build logic. For that very reason, it's easy to end up with a build script that mixes declarative DSL elements with imperative, procedural code. Let's talk about some concrete examples.
+
+* _Declarative code:_ Built-in, language-agnostic DSL elements (e.g. api:org.gradle.api.Project#dependencies(groovy.lang.Closure)[] or api:org.gradle.api.Project#repositories(groovy.lang.Closure)[]) or DSLs exposed by plugins
+* _Imperative code:_ Conditional logic or very complex task action implementations
+
+The end goal of every build script should be to only contain declarative language elements which makes the code easier to understand and maintain. Imperative logic should live in binary plugins and which in turn is applied to the build script. As a side product, you automatically enable your team to link:https://guides.gradle.org/designing-gradle-plugins/#reusable_logic_should_be_written_as_binary_plugin[reuse the plugin logic in other projects] if you publish the artifact to a binary repository.
+
+The following sample build shows a negative example of using conditional logic directly in the build script. While this code snippet is small, it is easy to imagine a full-blown build script using numerous procedural statements and the impact it would have on readability and maintainability. By moving the code into a class link:https://guides.gradle.org/testing-gradle-plugins/[testability] also becomes a valid option.
+
+++++
+<sample id="conditionalLogicDont" dir="userguide/bestPractices/conditionalLogic/dont" title="A build script using conditional logic to create a task">
+    <sourcefile file="build.gradle"/>
+</sample>
+++++
+
+Let's compare the build script with the same logic implemented as a binary plugin. The code might look more involved at first but clearly looks more like typical application code. This particular plugin class lives in the <<sec:build_sources,`buildSrc` directory>> which makes it available to the build script automatically.
```

Comment:
I like how Oehme says in [his blog post](https://oehme.github.io/use-buildSrc)

> A script is just a plugin without the class/method boilerplate

We probably can ~steal~ re-use a bunch of his words verbatim for this. They're good words.

Reply:
Thanks for the pointer. Wasn't aware of the blog post. I changes some things around.